hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java

```diff
@@ -968,6 +968,19 @@ private RegionLocations locateRegionInMeta(TableName tableName, byte[] row, bool
     }
   }
 
+  private void takeUserRegionLock() throws IOException {
+    try {
+      long waitTime = connectionConfig.getScannerTimeoutPeriod();
+      if (!userRegionLock.tryLock(waitTime, TimeUnit.MILLISECONDS)) {
```

Comment:
@saintstack Rushabh and I had a quick offline chat about this PR. We were wondering what is the right timeout to use for this lock. In the client code path there are a bunch of time out configurations depending on the path we take and sometimes layered on top of each other.

Specifically I was wondering if hbase.client.operation.timeout would be the right one to use for this. I understand that we are using the scanner timeout here because the call wraps a scanner with the same timeout. From a client standpoint though, scanner is just an implementation detail of locateRegion (root caller in this case) and that root caller should be wrapped with a general operation timeout rather than a timeout that is specific to the scanner.

Not a big deal but I was just curious and would like to know your thoughts. Thanks.

Reply:
> We were wondering what is the right timeout to use for this lock.

+1. The scanner timeout is not a good choice here.

Reply:
@bharathv  @infraio  Thank you for your feedback. 

> Specifically I was wondering if hbase.client.operation.timeout would be the right one to use for this. 

I don't feel hbase.client.operation.timeout is the right choice here too. This config is meant for the whole end to end operation timeout which includes all layers of retries and the default value is 20 mins. If we use this timeout then we are not gaining anything.

We can introduce a new config property (something like hbase.client.lock.timeout.period) and default it to something like 10 seconds. That way we don't depend on existing scanner/operation timeout periods. Let me know what you guys think. Thank you !

Reply:
This is really difficult to operate already, just look at us discussing it now. Scanner timeout? Operation timeout? Both! Neither! Make another! Let's not introduce another config option.

Reply:
Good concern @bharathv .

operation timeout is for 'whole operation end-to-end' per @shahrs87 . Here we are doing a sub-task so operation timeout doesn't seem right. Scanner timeout seems good; when it expires the scan will throw and we'll do the finally block anyways?

What would you suggest @infraio ?

I agree w/ @apurtell that last thing we need is new timeout ; client timeout is fraught as is.

Reply:
> operation timeout is for 'whole operation end-to-end'

In that case IMO the right fix is to extend the operation timeout to all public methods of HTable (may or may not be a part of this change). Looking at the code, we do it for some and we skip it for others. One of those methods is locateRegion() which is the root caller in this case. 

My concern was from a user configuration POV, there 3 or 4 different timeouts a user has to configure (depending on which API they are using) to get a proper timeout behavior. Ideally it'd be nice if there is a single timeout at the root level that is an e-e timeout and probably other timeouts like scanner timeout etc for finer control depending on the need.

Don't mean to block this PR, we can probably implement that as a separate change.

Reply:
`My concern was from a user configuration POV, there 3 or 4 different timeouts a user has to configure (depending on which API they are using) to get a proper timeout behavior. Ideally it'd be nice if there is a single timeout at the root level that is an e-e timeout and probably other timeouts like scanner timeout etc for finer control depending on the need.`

This is a critical concern. Our @petersomogyi was the last to throw himself against this rampart. The work is not yet finished. Yes, to a new issue.

Here will wait on @infraio 's input.....

Reply:
I am +1 to use operation timeout here.

Reply:
> I am +1 to use operation timeout here.

@infraio  the default value for operation timeout is 20 minutes. Do you want clients to wait for 20 minutes to acquire this lock ?

Reply:
@saintstack  @apurtell  @infraio  We need to come to consensus on which timeout property to use ? IMHO don't think we should use operation timeout.

Reply:
> the default value for operation timeout is 20 minutes. Do you want clients to wait for 20 minutes to acquire this lock ?

The problem is not how long the default value. The problem here is to use the right timeout. And for the default 20min operation timeout, we set it to 20mins because it is easy for passing ITBLL. And in real producetion cluster, no user will use the default value if it serve for online service. 

The guarantee of operation timeout is all table call (exclude scan) should return or throw exception when timeout. And when need to locate, locateRegionInMeta is one step of table call. So it should to keep this guarantee.

Reply:
Operation timout is not the best choice but better than scanner timeout. Thanks.

Reply:
We have an internal customer who wants a strict 15 seconds SLA for every operation. Since operation timeout is end to end timeout which includes all the retries, sleep within retries so we suggested them to use operation timeout as 15 seconds.
Also we recommended them to set scanner timeout period to 7 seconds with  retries config (hbase.client.retries.number)  set to 2 .  There is some sleep interval between each attempt and we expect the call to complete within 15 seconds and if it doesn't then operation timeout will kick in and fail the call.
But we found out that getRegionLocations call is not bounded by operation timeout.
Now if we set the lock timeout to same as operation timeout, then in worst case scenario call will fail in 15 (lock timeout) + 15 (lock timeout 2nd try) + 1 (assuming sleep of 1 second) = 31 seconds
If we set the lock timeout to same as scanner timeout, then in worst case scenario the call will fail in 7 (scanner timeout) + 7 (scanner timeout) + 1 (sleep between tries) = 16 seconds which is closer to SLA that we promised.
Hope this makes sense. If still the community wants to go forward with operation timeout, I will change it to operation timeout.
@infraio  @bharathv  @virajjasani  @saintstack  @apurtell

Reply:
I just had an offline chat with @shahrs87 and I am +1 with keeping scanner timeout as config for Lock, the intention is to:
1. not introduce more timeout configs and worsen timeout burden.
2. use the one which can help with multiple retries (if allowed within SLA) rather than timeout which is available at entire operation level (rpc call + lock + locate region by scan + retry if required + and so on...).

Reply:
Got your case. I thought the problem is if lock timeout, the 2nd retry need to consider this case and should not retry.

Let me give a example about the rpc timeout (configed as 4 secs) and operation timeout(configed as 10 secs). The rpc timeout maybe reset when the remaining time is less than rpc timeout.
1. rpc call with 4 secs timeout.
2. retry call with 4 secs timeout.
3. retry call with 2 secs timeout which equals 10 - 4 - 4.

For this case, we need same guarantee. If the remaining time is zero, it should not to retry. I thought this is the right way to fix this problem. Thanks.

Reply:
Given that we have 4 +1s on using hbase.client.scanner.timeout.period, can we commit this patch ?  Also I have created https://issues.apache.org/jira/browse/HBASE-24983 for wrapping scan operation under operation timeout. 
@apurtell  @saintstack  @bharathv  @virajjasani  @infraio  what do you guys think ?

Reply:
> then in worst case scenario call will fail in 15 (lock timeout) + 15 (lock timeout 2nd try) + 1 (assuming sleep of 1 second) = 31 seconds

Reread RpcRetryingCallerImpl#callWithRetries code. If callable.prepare throw the LockTimeoutException, it will not retry. There are a check about call duration and callTimeout in line 156. See https://github.com/apache/hbase/blob/branch-2/hbase-client/src/main/java/org/apache/hadoop/hbase/client/RpcRetryingCallerImpl.java#L156

Reply:
> If callable.prepare throw the LockTimeoutException, 

IIUC the code, callable.prepare will never throw LockTimeoutException. We throw LockTimeoutException [here](https://github.com/apache/hbase/blob/f3210c8908809fa154118454e6a23f57bc222db1/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java#L866)  At this place we haven't created ReversedClientScanner object also. There is one retry loop [here](https://github.com/apache/hbase/blob/f3210c8908809fa154118454e6a23f57bc222db1/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java#L848) which will retry if exception is not TNFE or retries are not exhausted.
@infraio  please correct me if I misunderstood anything.

Reply:
bq. There is one retry loop here which will retry if exception is not TNFE or retries are not exhausted.
Please check the code. The LockTimeoutException will be throwed out and terminate the loop. It is not in the try...catch code block.
bq. IIUC the code, callable.prepare will never throw LockTimeoutException.
callable.prepare will call locateRegionInMeta inside. So  locateRegionInMeta throw LockTimeoutException, then it will throw this exception out. 

Thanks.

Reply:
What is the resolution here @shahrs87  + @infraio ?

Reply:
> Please check the code. The LockTimeoutException will be throwed out and terminate the loop. It is not in the try...catch code block.

@infraio  you are right. Fixed that in latest commit. Please review again.

Reply:
> What is the resolution here @shahrs87 + @infraio ?

@infraio  pointed out one bug in my patch. Fixed it yesterday. Waiting for his feedback. Thank you for being so patient !

Reply:
@infraio  could you please review again ?